src/storage/src/dedup_pk_cell_based_row_serializer.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+use std::iter::Iterator;
+
+use risingwave_common::array::Row;
+use risingwave_common::catalog::{ColumnDesc, ColumnId};
+use risingwave_common::error::Result;
+use risingwave_common::types::VirtualNode;
+
+use crate::cell_based_row_serializer::CellBasedRowSerializer;
+use crate::cell_serializer::{CellSerializer, KeyBytes, ValueBytes};
+
+/// [`DedupPkCellBasedRowSerializer`] is identical to [`CellBasedRowSerializer`].
+/// Difference is that before serializing a row, pk datums are filtered out.
+pub struct DedupPkCellBasedRowSerializer {
+    /// Contains:
+    /// 1. Row indices of datums not in pk,
+    /// 2. or datums which have to be stored regardless
+    ///    (e.g. if memcomparable not equal to value encoding)
+    dedup_datum_indices: HashSet<usize>,
```

Comment:
I'm not sure whether consider a situation like: 

all cells of a state table is pk (table_columns.len = pk.len). just like #3474 . In this case, we can not dedup all pk, but at least one pk is left.

Reply:
Can `SENTINEL_CELL_ID` help with this case? Would require changes to the logic of deserialization though.

Reply:
Adding some tests to verify this

Reply:
Added: https://github.com/singularity-data/risingwave/blob/cc4d1d196313f9bebfa2fd44e2cd3436d70652b8/src/storage/src/dedup_pk_cell_based_row_serializer.rs#L200

Tests the case where all datums are dedupped. As @lmatz mentioned, `SENTINEL` cell will still be pushed:

https://github.com/singularity-data/risingwave/blob/6a2f4e3d918cf2af4a149d42c4e7e710879f9f26/src/common/src/util/ordered/serde.rs#L221-L223

Reply:
> Would require changes to the logic of deserialization though.

Regarding deserialization, we can also rely on `cell_based_row_deserializer` to do it, so I don't think much changes are needed:

https://github.com/singularity-data/risingwave/blob/cc4d1d196313f9bebfa2fd44e2cd3436d70652b8/src/storage/src/dedup_pk_cell_based_row_serializer.rs#L222-L243

Just need to replace the deduplicated datums with datums from pk. This logic can be in `dedup_pk_cell_based_row_deserializer`.